File: Lab-1/Example3/Deadlock.java
public class Deadlock {

    static final Object L1 = new Object();
    static final Object L2 = new Object();

    static void thread1() {
        synchronized (L1) {
            System.out.println("Thread 1 acquired L1");
            try { Thread.sleep(100); } catch (InterruptedException e) {}

            System.out.println("Thread 1 waiting for L2...");
            synchronized (L2) {
                System.out.println("Thread 1 acquired L2");
            }
        }
    }

    static void thread2() {
        synchronized (L2) {
            System.out.println("Thread 2 acquired L2");
            try { Thread.sleep(100); } catch (InterruptedException e) {}

            System.out.println("Thread 2 waiting for L1...");
            synchronized (L1) {
                System.out.println("Thread 2 acquired L1");
            }
        }
    }

    public static void main(String[] args) throws InterruptedException {
        Thread t1 = new Thread(Deadlock::thread1);
        Thread t2 = new Thread(Deadlock::thread2);

        t1.start();
        t2.start();

        t1.join();
        t2.join();

        System.out.println("Both threads finished (should never print!)");
    }
}

File: Lab-1/Example3/Deadlock.cpp
#include <iostream>
#include <thread>
#include <mutex>
#include <chrono>

std::mutex L1, L2;

void thread1() {
    std::lock_guard<std::mutex> lock1(L1);
    std::cout << "Thread 1 acquired L1" << std::endl;

    std::this_thread::sleep_for(std::chrono::milliseconds(100)); // deadlock'u görünür kılmak için

    std::cout << "Thread 1 waiting for L2..." << std::endl;
    std::lock_guard<std::mutex> lock2(L2);
    std::cout << "Thread 1 acquired L2" << std::endl;
}

void thread2() {
    std::lock_guard<std::mutex> lock2(L2);
    std::cout << "Thread 2 acquired L2" << std::endl;

    std::this_thread::sleep_for(std::chrono::milliseconds(100));

    std::cout << "Thread 2 waiting for L1..." << std::endl;
    std::lock_guard<std::mutex> lock1(L1);
    std::cout << "Thread 2 acquired L1" << std::endl;
}

int main() {
    std::thread t1(thread1);
    std::thread t2(thread2);

    t1.join();
    t2.join();

    std::cout << "Both threads finished (this line should never print!)" << std::endl;
    return 0;
}
